### ISSUE TITLE
Async callbacks are not actually async?

### ISSUE THREAD

**Issue by SachaG** [VulcanJS/Vulcan#2205]
It seems like async callbacks (at least, collection callbacks in `mutators.js`) are holding back mutations from returning their value to the client, when their whole point is that they shouldn't affect the mutation at all.

**Comment by jimrandomh** [comment]
Whether `runCallbacksAsync` starts things up in the background, or is async/await-style and blocks you if you await it, is weirdly ambiguous. It looks like [this commit](https://github.com/VulcanJS/Vulcan/commit/87e3721f34d02adaf869e7801844167f10be8ca0#diff-8683712f666fad3e3ad2602a543821c5) switched it from the former to the latter. But the calls to `runCallbacksAsync` in `mutators.js` already had `await`s on them, which had no effect, implying preexisting confusion about just what `runCallbacksAsync` was supposed to do.

The old start-things-in-the-background behavior of `runCallbacksAsync` was actually untenable enough that LessWrong rewrote that function in our fork. The problem is that we use it mainly for filling in denormalized fields in the database, and lauching things in the background means that unit tests can't rely on them having finished. Our [modified version](https://github.com/LessWrong2/Vulcan/blob/master/packages/vulcan-lib/lib/modules/callbacks.js) keeps track of which callbacks are outstanding, and adds a function `waitUntilCallbacksFinished` for unit tests to use.

**Comment by SachaG** [comment]
Yeah `waitUntilCallbacksFinished` sounds like a good solution. So I assume that if you have that running things in the background becomes less of a problem? Because it's not great UX to have the user wait until notifications, etc. have been sent after submitting a form.

**Comment by jimrandomh** [comment]
I think the new behavior of `runCallbacksAsync` is better; if mutators want the callbacks to happen in the background, the way to do that should be to leave off the `await`. That way other users of the callbacks API (if they exist) can `await` it if they want.

**Comment by jimrandomh** [comment]
(Right now `runCallbacks` leads to a dependency on Meteor fibers, which would be nice to cut down on, since if you need to do a database query inside a callback you aren't async and can't use `await`. It would be nice to make all the async-ness explicit, by moving away from `runCallbacks` and towards `runCallbacksAsync`, making the `runCallbacks` users async themselves where possible.)

**Comment by SachaG** [comment]
Maybe we should involve @ochicf in this discussion since he's the one who worked on callbacks last I think. 

Personally I'm happy to go in any direction as long as both use cases (wait for result/don't wait for result) are supported. But I think the tricky part is going to be naming, since now we're basically saying that using `async/await` will make a callback wait, which is what we called "sync" before…

**Comment by ochicf** [comment]
Thanks for pinging me and sorry for the 3 days delay.

This topic has already been discussed in [#1999](https://github.com/VulcanJS/Vulcan/issues/1999) and specially [#1949](https://github.com/VulcanJS/Vulcan/issues/1949).

The main issue I see with the entire hook/callback system derives from naming and mixed concepts. Right now Vulcan is using `sync` and `async` and I think that it does not specify enough how the callbacks are run, and that extends to:

1. how the callbacks are run between them?
2. what result and which shape will provide? Or more splicit: how calling multiple callbacks upon a value will affect on its outcome"?
3. will the caller code wait for the result or not?

In my opinion, this are the answers to the previous questions for the current `runCallbacks` and `runCallbacksAsync`:

1. `runCallbacks` runs callbacks as reducers, so it waits between callbacks and passes the outcome of one as the input to the next. `runCallbacksAsync` runs callbacks in parallel, so they will all have the same input and there may be race conditions (think of this as running promises with `Promise.all`).
2. `runCallbacks` returns a modified version of the input, `runCallbacksAsync` returns as many modified versions of the input (one for each callback run). This is why it is optional to return a value (and also why usually it is not used by the caller code).
3. both `runCallbacks` and `runCallbacksAsync` can wait for results or not, and each of the callback may choose if its result is awaited or detached

Now, the last point is the one relevant to this issue. Note that each callback is able to detach itself from the caller code. This is more useful for `runCallbacksAsync`, that is probably being used to perform side effects.

I've written a simple snippet that shows how to detach from the caller execution:

```js

import { addCallback, runCallbacksAsync } from 'meteor/vulcan:core';

const sleep = time =>
  new Promise(resolve => {
    setTimeout(resolve, time);
  });

addCallback('some.async.hook', async function doWait() {
  // awaiting directly on the callback, hence returning a promise that will
  // be awaited
  console.log('doWait.before');
  await sleep(2000);
  console.log('doWait.after');
  return `did wait`;
});

addCallback('some.async.hook', function dontWait() {
  // note that we are not returing a promise, so we are detaching from the
  // awaited code
  console.log('dontWait.before');
  setTimeout(async () => {
    // here you can perform your expensive computational side effects without
    // having the caller waiting!
    console.log('dontWait.async.before');
    await sleep(10000);
    console.log('dontWait.async.after');
  });
  console.log('dontWait.after');
  return `did not wait`;
});

const test = async () => {
  console.log('starting test');
  console.time('test');
  const results = await runCallbacksAsync({
    name: 'some.async.hook',
  });
  console.timeEnd('test');
  console.log('test results:', results);
};

test();
```

This are the logs produced (notice the timestamps):

```
I20190303-17:49:39.858(1)? starting test
I20190303-17:49:40.082(1)? doWait.before
I20190303-17:49:40.082(1)? dontWait.before
I20190303-17:49:40.082(1)? dontWait.after
I20190303-17:49:40.153(1)? dontWait.async.before
I20190303-17:49:42.084(1)? doWait.after
I20190303-17:49:42.097(1)? test: 2225.800ms
I20190303-17:49:42.098(1)? test results: [ 'did wait', 'did not wait' ]
I20190303-17:49:50.157(1)? dontWait.async.after
```

Conclusion:

- caller code can choose to wait or not (with `await` when running `runCallbacksAsync()`). I would recommend to always `await`
- each callback can choose if the caller will await on it, or if it calls in a detached manner by returning a promise or not

I hope that it makes sense. Let me know if it doesn't or if something was not clear enough.

**Comment by SachaG** [comment]
@ochicf I'll be honest, it's been a bit hard for me to keep up with this issue as I'm currently focusing on other parts of Vulcan… So if you're ok with this I would just give you the lead on this and let you implement your suggestions. 

Ideally there would be no breaking changes, but if there are this is probably a good time to do it since the next update includes breaking changes anyway.
